### PR TITLE
Fixes a bug when arrays are decoded to an interface

### PR DESCRIPTION
**Pull request by jaekwon**
Two gotchas found:

* If you don't use `defer mtx.Unlock()`, then the state of mutex usage may become inconsistent if you panic out.  We are panic'ing correctly, but to make the API most useful, this PR makes some extra work to make the state usable even after recovering from a panic.  In general this doesn't seem like something we can get right 100% of the time, but it seems useful to do when it's easy enough.

* After you call rv.Set(crv) where rv.Kind() == reflect.Interface and crv.Kind() == reflect.Array, then even if crv.CanAddr(), mutating crv won't mutate rv because an array more or less acts as a single primitive.  You can take a slice of an addressable array, but once you copy (e.g. rv.Set) it somewhere, now you have two independent slices.

### CODE REVIEW

**Comment by jaekwon** [comment]
Closing as it most assuredly fixes the bug, but creating a PR for discussion.

File: codec.go [hunk 1]
```diff
@@ -136,20 +141,22 @@ func (cdc *Codec) RegisterConcrete(o interface{}, name string, opts *ConcreteOpt
 	var info = cdc.newTypeInfoFromConcreteType(rt, pointerPreferred, name, opts)
 
 	// Finally, check conflicts and register.
-	cdc.mtx.Lock()
-	{
+	func() {
```
Comment (odeke-em) [inline codec.go:144]: Perhaps this anonymous closure was unnecessary? Why not just
```go
cdc.mtx.Lock()
defer cdc.mtx.Unlock()

cdc.addCheckConflictsWithConcrete_nolock(info)
cdc.setTypeInfo_nolock(info)
```

File: codec.go [hunk 2]
```diff
@@ -98,13 +99,17 @@ func (cdc *Codec) RegisterInterface(ptr interface{}, opts *InterfaceOptions) {
 	var info = cdc.newTypeInfoFromInterfaceType(rt, opts)
 
 	// Finally, check conflicts and register.
-	cdc.mtx.Lock()
-	{
+	func() {
```
Comment (odeke-em) [inline codec.go:102]: Same here about this closure being unnecessary, we could have just added `defer cdc.mtx.Unlock()`

File: codec.go [hunk 3]
```diff
@@ -420,11 +429,52 @@ func (cdc *Codec) addCheckConflictsWithConcrete_nolock(cinfo *TypeInfo) {
...
+}
+
+//----------------------------------------
+// .String()
+
+func (ti TypeInfo) String() string {
+	buf := new(bytes.Buffer)
+	buf.Write([]byte("TypeInfo{"))
+	buf.Write([]byte(fmt.Sprintf("Type:%v,", ti.Type)))
+	if ti.Type.Kind() == reflect.Interface {
```
Comment (odeke-em) [inline codec.go:453]: In some of these spots we can just use
```go
fmt.Fprintf(buf, "
```

**Pull request merged**